### ISSUE TITLE
tracker 

### ISSUE THREAD

**Issue by 3073** [ahmetozlu/vehicle_counting_tensorflow#1]
why tracking algorithm is not required?  
I am to count the number of people entering a vehicle.   How can I provide the ROI line so that people leaving or entering past that line will be counted?

thank you

**Comment by ahmetozlu** [comment]
The idea is simple:

The input video or real time camera record is read frame by frame via OpenCV.

The each frame will be send to TensorFlow Object Detection Model as an input. TensorFlow detects the objects on each frame. This process is repeated again and again till the video finish and so while frames which has bounding boxes on detected object are reading, processed and showing consecutively so objects are being tracked and you do not need any special tracker for it.

Your project topic (counting the number of people entering a vehicle) sounds interesting! You should change the ROI Line position with suitable angle and TensorFlow detect the people entering or exiting vehicle and also you can count them by this way.

Just focus to change ROI line angle and remove unnecessary modules sucs as color recognition module or speed and direction prediction module.

The ROI Line is assigned on line 252 in [this class](https://github.com/ahmetozlu/vehicle_counting_tensorflow/blob/master/utils/visualization_utils.py#L252). Focus on it to change the line with true angle and do some modifications on general code to fit the program on your case.

**Comment by 3073** [comment]
thank you. I will try it and post the result

**Comment by delonleo** [comment]
Interesting way for counting vehicle, I must say.

Can I understand line 252 in [this class](https://github.com/ahmetozlu/vehicle_counting_tensorflow/blob/master/utils/visualization_utils.py#L252) ?  How do you change the ROI Line angle ?

What do the code below means ? What if i change 205 or 210 ?

`if ((len(bottom_position_of_detected_vehicle_temp_list) != 0) and bottom-bottom_position_of_detected_vehicle_temp_list[0] > 0 and 205 < bottom_position_of_detected_vehicle_temp_list[0] and bottom_position_of_detected_vehicle_temp_list[0]< 210 and 200 < bottom ):`

**Comment by ahmetozlu** [comment]
`if ((len(bottom_position_of_detected_vehicle_temp_list) != 0) and bottom-bottom_position_of_detected_vehicle_temp_list[0] > 0 and 205 < bottom_position_of_detected_vehicle_temp_list[0] and bottom_position_of_detected_vehicle_temp_list[0]< 210 and 200 < bottom ):`

This means that, if the detected vehicle's bottom y pixel values (actually, bounding box's bottom y pixel values) is between 205 and 210, increase the counter. Thus, you can increase by 1 the detected vehicle count.

If you want to change ROI line, you should play with these pixel value range. For example, if you want to move above the ROI line as y axis, just set that range 

`185 < bottom_position_of_detected_vehicle_temp_list[0] and bottom_position_of_detected_vehicle_temp_list[0]< 200`

Maybe it looks a bit complicated but the developing is on progress, these points will be divided more functions are in different classes so it will be easier to understand and manipulate.

**Comment by SahibYar** [comment]
Thanks for your explanation, can you please explain how you are calculating the speed [in this block](https://github.com/ahmetozlu/vehicle_counting_tensorflow/blob/master/utils/visualization_utils.py#L266``) ?
Because I can't understand some of the transformations like `scale_real_length`, `scale_real_time_passed`, `speed = speed/6*40` and `scale = 1`
**please elaborate a bit.**

**Comment by Gerry-K** [comment]
Ahmetozlu, great work!! One thing we're struggeling with is the counting part. If the object recognition loses track (for a very short time), there's no bounding box aound the object thus doesn't register once it passes the ROI line. Widening the ROI line caused vehicles being counted twice because the bottom of the bounding box hits the ROI line, tracking is lost and regained agai, hitting the RIO line again.

**Comment by ahmetozlu** [comment]
@Gerry-K , 

Widening ROI Line causes counting one vehicle as 2-3 or more maybe. You should consider about [fine-tuning](https://www.youtube.com/watch?v=H8sXcAXrGR4) the [coco ](http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_coco_2017_11_17.tar.gz) model that is used in this work so you can detect vehicles in your case with higher accuracy (you can boost the accuracy via fine-tuning). Thus, your object recognition will not loses track!

If you will make wider ROI Line, then you should consider about the counting range so you should manipulate this [line](https://github.com/ahmetozlu/vehicle_counting_tensorflow/blob/master/utils/visualization_utils.py#L252)

**Comment by 3073** [comment]
@ahmetozlu :  there is no increase of the bounding box in the Y pixel. the change in bounding box pixel is along the X axis. since the vehicle is moving in the x direction. why we consider the y pixel? that was logical if the object were moving upstairs or downstairs ( it's y coordinate decrease and increase based on the situation)

**Comment by ahmetozlu** [comment]
The pipeline and explanation of the tracker are available in [here](https://github.com/ahmetozlu/vehicle_counting_tensorflow#tracker)

I have developed my system just considering y pixel because it was enough for my case. However, you can adopt it for your case, you can use x pixel ROI line and adopt this project to count, predict speed and color of the vehicles that are on moving on x direction, moreover you can adopt it for the vehicles are moving on both x and y angles.
